File: Source/ARTHttp.m
#import "ARTHttp+Private.h"
#import "ARTURLSessionServerTrust.h"
#import "ARTConstants.h"
#import "ARTInternalLog.h"

@interface ARTHttp ()

@property (readonly, nonatomic) id<ARTURLSession> urlSession;

@end

Class configuredUrlSessionClass = nil;

#pragma mark - ARTHttp

@implementation ARTHttp {
    ARTInternalLog *_logger;
}

+ (void)setURLSessionClass:(const Class)urlSessionClass {
    configuredUrlSessionClass = urlSessionClass;
}

- (instancetype)initWithQueue:(dispatch_queue_t)queue logger:(ARTInternalLog *)logger {
    self = [super init];
    if (self) {
        const Class urlSessionClass = configuredUrlSessionClass ? configuredUrlSessionClass : [ARTURLSessionServerTrust class];
        _urlSession = [[urlSessionClass alloc] init:queue];
        _logger = logger;
    }
    return self;
}

- (ARTInternalLog *)logger {
    return _logger;
}

- (dispatch_queue_t)queue {
    return _urlSession.queue;
}

- (void)dealloc {
    [_urlSession finishTasksAndInvalidate];
}

- (NSObject<ARTCancellable> *)executeRequest:(NSMutableURLRequest *)request completion:(ARTURLRequestCallback)callback {
    ARTLogDebug(self.logger, @"--> %@ %@\n  Body: %@\n  Headers: %@", request.HTTPMethod, request.URL.absoluteString, [self debugDescriptionOfBodyWithData:request.HTTPBody], request.allHTTPHeaderFields);

    return [_urlSession get:request completion:^(NSHTTPURLResponse *response, NSData *data, NSError *error) {
        NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse *)response;
        if (error) {
            ARTLogError(self.logger, @"<-- %@ %@: error %@", request.HTTPMethod, request.URL.absoluteString, error);
        } else {
            ARTLogDebug(self.logger, @"<-- %@ %@: statusCode %ld\n  Data: %@\n  Headers: %@\n", request.HTTPMethod, request.URL.absoluteString, (long)httpResponse.statusCode, [self debugDescriptionOfBodyWithData:data], httpResponse.allHeaderFields);
            NSString *headerErrorMessage = httpResponse.allHeaderFields[ARTHttpHeaderFieldErrorMessageKey];
            if (headerErrorMessage && ![headerErrorMessage isEqualToString:@""]) {
                ARTLogWarn(self.logger, @"%@", headerErrorMessage);
            }
        }
        callback(httpResponse, data, error);
    }];
}

- (NSString *)debugDescriptionOfBodyWithData:(NSData *)data {
    if (self.logger.logLevel <= ARTLogLevelDebug) {
        NSString *requestBodyStr = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
        if (!requestBodyStr) {
            requestBodyStr = [data base64EncodedStringWithOptions:NSDataBase64EncodingEndLineWithCarriageReturn];
        }
        return requestBodyStr;
    }
    return nil;
}

@end

File: Sources/AblySwift/ARTHttp.swift
import Foundation

// swift-migration: original location ARTHttp.h, line 14
/// :nodoc:
internal protocol ARTHTTPExecutor {
    // swift-migration: original location ARTHttp.h, line 16
    func executeRequest(_ request: URLRequest, completion: ARTURLRequestCallback?) -> (any ARTCancellable)?
}

// swift-migration: original location ARTHttp.m, line 12
private var configuredUrlSessionClass: AnyClass? = nil

// swift-migration: original location ARTHttp.h, line 22 and ARTHttp.m, line 16
/// :nodoc:
internal class ARTHttp: NSObject, ARTHTTPExecutor {
    
    // swift-migration: original location ARTHttp.m, line 8
    internal let urlSession: any ARTURLSession
    
    // swift-migration: original location ARTHttp.m, line 17
    private let _logger: ARTInternalLog
    
    // swift-migration: original location ARTHttp.h, line 24 and ARTHttp.m, line 20
    internal class func setURLSessionClass(_ urlSessionClass: AnyClass) {
        configuredUrlSessionClass = urlSessionClass
    }
    
    // swift-migration: original location ARTHttp.h, line 27 and ARTHttp.m, line 24
    internal init(queue: DispatchQueue, logger: ARTInternalLog) {
        let urlSessionClass = configuredUrlSessionClass ?? ARTURLSessionServerTrust.self
        self.urlSession = (urlSessionClass as! ARTURLSession.Type).init(queue)
        self._logger = logger
        super.init()
    }
    
    // swift-migration: original location ARTHttp.m, line 34
    internal var logger: ARTInternalLog {
        return _logger
    }
    
    // swift-migration: original location ARTHttp+Private.h, line 7 and ARTHttp.m, line 38
    internal var queue: DispatchQueue {
        return urlSession.queue
    }
    
    // swift-migration: original location ARTHttp.m, line 42
    deinit {
        urlSession.finishTasksAndInvalidate()
    }
    
    // swift-migration: original location ARTHttp.h, line 16 and ARTHttp.m, line 46
    internal func executeRequest(_ request: URLRequest, completion callback: ARTURLRequestCallback?) -> (any ARTCancellable)? {
        let bodyString = debugDescriptionOfBody(with: request.httpBody)
        ARTLogDebug(self.logger, "--> \(request.httpMethod ?? "") \(request.url?.absoluteString ?? "")\n  Body: \(bodyString ?? "")\n  Headers: \(request.allHTTPHeaderFields ?? [:])")
        
        return urlSession.get(request) { [weak self] response, data, error in
            guard let self = self else { return }
            
            let httpResponse = response
            if let error = error {
                ARTLogError(self.logger, "<-- \(request.httpMethod ?? "") \(request.url?.absoluteString ?? ""): error \(error)")
            } else if let httpResponse = httpResponse {
                let dataString = self.debugDescriptionOfBody(with: data)
                ARTLogDebug(self.logger, "<-- \(request.httpMethod ?? "") \(request.url?.absoluteString ?? ""): statusCode \(httpResponse.statusCode)\n  Data: \(dataString ?? "")\n  Headers: \(httpResponse.allHeaderFields)\n")
                let headerErrorMessage = httpResponse.allHeaderFields[ARTHttpHeaderFieldErrorMessageKey] as? String
                if let headerErrorMessage = headerErrorMessage, !headerErrorMessage.isEmpty {
                    ARTLogWarn(self.logger, "\(headerErrorMessage)")
                }
            }
            callback?(httpResponse, data, error)
        }
    }
    
    // swift-migration: original location ARTHttp.m, line 64
    private func debugDescriptionOfBody(with data: Data?) -> String? {
        guard let data = data else { return nil }
        
        if self.logger.logLevel.rawValue <= ARTLogLevel.debug.rawValue {
            var requestBodyStr = String(data: data, encoding: .utf8)
            if requestBodyStr == nil {
                requestBodyStr = data.base64EncodedString(options: .lineLength76Characters)
            }
            return requestBodyStr
        }
        return nil
    }
}